components/concurrency_manager/src/metrics.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! Metrics for KeyHandle memory leak detection and monitoring.
+
+use lazy_static::lazy_static;
+use prometheus::{
+    register_int_counter, register_int_counter_vec, register_int_gauge, register_int_gauge_vec,
+    IntCounter, IntCounterVec, IntGauge, IntGaugeVec,
+};
+
+lazy_static! {
+    /// Gauge for current KeyHandle instances, labeled by key type (data/meta)
+    pub static ref KEYHANDLE_CURRENT: IntGaugeVec = register_int_gauge_vec!(
+        "tikv_keyhandle_current",
+        "Current number of KeyHandle instances in memory",
+        &["type"]
+    )
+    .unwrap();
+
+    /// Counter for KeyHandle creation events, labeled by key type (data/meta)
+    pub static ref KEYHANDLE_CREATED_TOTAL: IntCounterVec = register_int_counter_vec!(
+        "tikv_keyhandle_created_total",
+        "Total number of KeyHandle instances created",
+        &["type"]
+    )
+    .unwrap();
+
+    /// Counter for KeyHandle destruction events, labeled by key type (data/meta)
+    pub static ref KEYHANDLE_RELEASED_TOTAL: IntCounterVec = register_int_counter_vec!(
+        "tikv_keyhandle_released_total",
+        "Total number of KeyHandle instances released",
+        &["type"]
+    )
+    .unwrap();
+
+    /// Counter for KeyHandleGuard acquisition events, labeled by key type (data/meta)
+    pub static ref KEYHANDLE_GUARD_ACQUIRED_TOTAL: IntCounterVec = register_int_counter_vec!(
+        "tikv_keyhandle_guard_acquired_total",
+        "Total number of KeyHandleGuard instances acquired",
+        &["type"]
+    )
+    .unwrap();
+
+    /// Counter for KeyHandleGuard release events, labeled by key type (data/meta)
+    pub static ref KEYHANDLE_GUARD_RELEASED_TOTAL: IntCounterVec = register_int_counter_vec!(
+        "tikv_keyhandle_guard_released_total",
+        "Total number of KeyHandleGuard instances released",
+        &["type"]
+    )
+    .unwrap();
+
+    /// Counter for key-path lifecycle checkpoints, labeled by path and checkpoint
+    pub static ref KEYHANDLE_CHECKPOINT_TOTAL: IntCounterVec = register_int_counter_vec!(
+        "tikv_keyhandle_checkpoint_total",
+        "Total number of KeyHandle lifecycle checkpoints reached",
+        &["path", "checkpoint"]
+    )
+    .unwrap();
+
+    // Pre-created checkpoint counters for performance optimization
+    pub static ref CHECKPOINT_PREWRITE_START: IntCounter =
```

Comment:
NIT: we can use `make_auto_flush_static_metric` to avoid this kind of boilerplate code.